src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -164,21 +164,34 @@ partitions_request_iterator validate_partitions(
       begin,
       valid_partitions_end,
       std::back_inserter(invalid_partitions),
-      error_code::invalid_replication_factor,
+      error_code::unknown_topic_or_partition,
       "Number of replicas does not match the topic replication factor",
-      [&tp_metadata](const reassignable_partition& partition) {
+      [&tp_metadata, tp_name{topic_response.name()}](
+        const reassignable_partition& partition) {
           if (!partition.replicas.has_value()) {
               return true;
           }
 
-          if (
-            !tp_metadata.has_value()
-            || !tp_metadata.value().is_topic_replicable()) {
+          if (!tp_metadata.has_value()) {
+              vlog(
+                klog.debug, "Topic metadata is undefined: topic {}", tp_name);
+              return false;
+          }
+
+          if (!tp_metadata.value().is_topic_replicable()) {
+              vlog(klog.debug, "Topic metadata is not replicable");
               return false;
           }
 
           auto tp_replication_factor
             = tp_metadata.value().get_replication_factor();
+          vlog(
+            klog.debug,
+            "Topic replication factor: topic {}, replication factor {}, "
+            "requested replicas {}",
+            tp_name,
+            tp_replication_factor,
+            *partition.replicas);
```

Comment:
This validator returns `unknown_topic_or_partition` in the final case replication factor doesn't match. What we need are two seperate validators, one for validating tp_metadata != null, then if that passes we can assert on the replication factor being equivalent to the request. That is the validation kafka would perform as well.

Reply:
There should be tests to check this condition, they should have failed with this change.